auth/iam_ipa.go

```diff
@@ -0,0 +1,394 @@
+	if err != nil {
+		return Account{}, err
+	}
+
+	uid, _ := strconv.Atoi(userResult.Uidnumber[0])
+	gid, _ := strconv.Atoi(userResult.Gidnumber[0])
+	account := Account{
+		Access:  access,
+		Role:    RoleUser,
+		UserID:  uid,
+		GroupID: gid,
+	}
+
+	session_key := make([]byte, 16)
+	rand.Read(session_key)
+	encrypted_key, err := rsa.EncryptPKCS1v15(rand.Reader, ipa.kraTransportKey, session_key)
+	if err != nil {
+		return account, err
+	}
+	req = ipa.newRequest("vault_retrieve_internal/1", []string{ipa.vaultName},
+		map[string]any{"username": access,
+			"session_key":   Base64EncodedWrapped(encrypted_key),
+			"wrapping_algo": "aes-128-cbc"})
+	data := struct {
+		Vault_data Base64EncodedWrapped
+		Nonce      Base64EncodedWrapped
+	}{}
+	_, err = ipa.rpc(req, &data)
+	if err != nil {
+		return account, err
+	}
+
+	aes, _ := aes.NewCipher(session_key)
+	cbc := cipher.NewCBCDecrypter(aes, data.Nonce)
+	cbc.CryptBlocks(data.Vault_data, data.Vault_data)
+	secret_unpadded_json, _ := pkcs7Unpad(data.Vault_data, 16)
+
+	secret := struct {
+		Data Base64Encoded
+	}{}
+	json.Unmarshal(secret_unpadded_json, &secret)
+	account.Secret = string(secret.Data)
+
+	fmt.Printf("%v\n", account)
+	return account, nil
+}
+
+func (ipa *ipaIAMService) UpdateUserAccount(access string, props MutableProps) error {
+	return fmt.Errorf("not implemented")
+}
+
+func (ipa *ipaIAMService) DeleteUserAccount(access string) error {
+	return fmt.Errorf("not implemented")
+}
+
+func (ipa *ipaIAMService) ListUserAccounts() ([]Account, error) {
+	return []Account{}, fmt.Errorf("not implemented")
+}
+
+func (ipa *ipaIAMService) Shutdown() error {
+	return nil
+}
+
+// Implementation
+
+func (ipa *ipaIAMService) login() error {
+	form := url.Values{}
+	form.Set("user", ipa.username)
+	form.Set("password", ipa.password)
+
+	req, err := http.NewRequest(
+		"POST",
+		fmt.Sprintf("%s/ipa/session/login_password", ipa.host),
+		strings.NewReader(form.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("referer", fmt.Sprintf("%s/ipa", ipa.host))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := ipa.client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == 401 {
+		return errors.New("cannot login to FreeIPA: invalid credentials")
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("cannot login to FreeIPA: status code %d", resp.StatusCode)
+	}
+	return nil
+}
+
+type rpcRequest = string
+
+type rpcResponse struct {
+	Result    json.RawMessage
+	Principal string
+	Id        int
+	Version   string
+}
+
+func (p rpcResponse) String() string {
+	return string(p.Result)
+}
+
+var errRpc = errors.New("IPA RPC error")
+
+func (ipa *ipaIAMService) rpc(req rpcRequest, value any) (rpcResponse, error) {
+
+	err := ipa.login()
+	if err != nil {
+		return rpcResponse{}, err
+	}
+
+	res, err := ipa.rpcInternal(req)
+	if err != nil {
+		return res, err
+	}
+	err = json.Unmarshal(res.Result, value)
```

Comment:
is this correct that the res.Result should be whatever is provided to the rpc method? I would have expected this should come from the response?

Reply:
Full rpc response isn't really used anywhere (it includes things like ipa version etc.). I left it in rpcInternal signature in case it is needed in the future and removed it from rpc(...) signature for now.